crates/hir-ty/src/infer.rs

```diff
@@ -699,7 +699,7 @@ impl<'a> InferenceContext<'a> {
             mismatch.expected = table.resolve_completely(mismatch.expected.clone());
             mismatch.actual = table.resolve_completely(mismatch.actual.clone());
             chalk_ir::zip::Zip::zip_with(
-                &mut UnknownMismatch(self.db),
+                &mut UnknownMismatch(self.db, |ty| matches!(ty.kind(Interner), TyKind::Error)),
```

Comment:
Not sure, but it feels like we should be checking here for projections?

Reply:
I have tried doing this at first time, but it seems to cause some errors in MIR (as shown below). I suspect that MIR uses `type_mismatch` in `borrowck` to avoid certain paths, so I only removed the corresponding error in the outer layer. However, I haven't studied this part carefully yet 🙊. If necessary, I can further investigate this error in MIR to ensure it's not another potential bug.

```
thread 'Worker' panicked at crates/hir-ty/src/mir.rs:185:21:
internal error: entered unreachable code: Only tuple has tuple field
```

Reply:
No, you're probably right, MIR is not resilient to code that doesn't typecheck.

Reply:
We should probably add a flag the type check results whether there were errors or other problems and have mir rely on that for exiting out early. Since us stripping diagnostics could cause MIR lowering to try lowering an non-successfully typed body.

Reply:
Agree. I'll try to implement it.

Reply:
I added an `ignore_in_diagnostics` field to the `TypeMismatch` struct, enabling us to retain type-mismatches for MIR lowering while removing them from diagnostics. It's more elegant than the previous solution.

Reply:
Hmmm, I found that after adding the field, we need to modify some code about tests and statistics to filter out ignored diagnostics. (done)